### ISSUE TITLE
another problem about multi-gpu 

### ISSUE THREAD

**Issue by cs-heibao** [NVIDIA/DALI#2158]
the main code are follows:
```
parser = argparse.ArgumentParser(description='Train cifar10 use DALI data process based on the resnet18')
parser.add_argument('--lr', default=0.1, type=float, help='learning rate')
parser.add_argument('--TRAIN_BS', default=4, type=int, help='batch size of data')
parser.add_argument('--GPUS', default=2, type=int, help='number of gpus')
parser.add_argument('--NUM_WORKERS', default=4, type=int)
parser.add_argument('--IMG_DIR', default='/*/DALI_process/ImageNet2012', type=str, help='data path')
parser.add_argument('--CROP_SIZE', default=224, type=int)
parser.add_argument('--IMAGENET_IMAGES_NUM_TRAIN', default=12724, type=int)

# multi-gpus
single_batch_size = 4

# Data
print('==> Preparing data..')
pip_train = [HybridTrainPipe(batch_size=single_batch_size, num_threads=args.NUM_WORKERS, device_id=device_id, data_dir=args.IMG_DIR+'/train', crop=args.CROP_SIZE, world_size=args.GPUS, local_rank=0) for device_id in range(args.GPUS)]
train_loader = DALIDataloader(pipeline=pip_train, size=args.IMAGENET_IMAGES_NUM_TRAIN, batch_size=args.TRAIN_BS, onehot_label=True)
start_time = time()
for batch_idx, (inputs, targets) in enumerate(train_loader):
    inputs = inputs.cuda()
    targets = targets.cuda()
    optimizer.zero_grad()
    outputs = net(inputs)
    loss = criterion(outputs, targets)
    loss.backward()
    optimizer.step()
print("INFO: Each epoch takes %s"%(time()- start_time))
```
my total images are 12724, and use 2 gpus, the total batchsize is 8 and each single_batch_size is 4. Here the len(train_loader) is 3181, and the data in train_loader is a list which has two elements, and each elements is a dict like {'data': *****, 'label': ****}, and the shape of inputs is (4, 3, 224, 224).
![image](https://user-images.githubusercontent.com/33689425/88644248-55024b00-d0f5-11ea-88e9-d6d852ac61d9.png)

but the problem is that, each epoch end with the iteration of 1591(12724/8=1590.5)
![image](https://user-images.githubusercontent.com/33689425/88644668-d6f27400-d0f5-11ea-85cd-29218c3401f0.png)

**Comment by JanuszL** [comment]
Hi,
Can you state what is the problem?

**Comment by cs-heibao** [comment]
@JanuszL 
my total images are 12724, and use 2 gpus, the total batchsize is 8 and each single_batch_size is 4. Here the len(train_loader) is 3181, and the data in train_loader is a list which has two elements, and each elements is a dict like {'data': *****, 'label': ****}, and the shape of inputs is (4, 3, 224, 224).
![image](https://user-images.githubusercontent.com/33689425/88644848-0903d600-d0f6-11ea-99ce-707a1e4094fd.png)
but the problem is that, each epoch end with the iteration of 1591(12724/8=1590.5)
![image](https://user-images.githubusercontent.com/33689425/88644916-1caf3c80-d0f6-11ea-87be-731c86ad7b83.png)

**Comment by cs-heibao** [comment]
@JanuszL 
while I find that the inputs always equal to train_loader._data_batches[0], can not read the train_loader._data_batches[1]

**Comment by JanuszL** [comment]
Hi,
DALI always works on the full batch size. When data from one shard is exhausted it just starts reading from the next one. If you want you may set the `pad_last_batch` reader option to True to clone the last sample in the batch instead of crossing the shard boundary. If you are using ExternalSource you need to do it on your own.
Is `DALIDataloader` some incarnation of `DALIGenericIterator` or your custom code?

**Comment by cs-heibao** [comment]
@JanuszL 
yes
```
from nvidia.dali.plugin.pytorch import DALIGenericIterator
from nvidia.dali.plugin.pytorch import DALIClassificationIterator

class DALIDataloader(DALIGenericIterator):
    def __init__(self, pipeline, size, batch_size, output_map=["data", "label"], auto_reset=True, onehot_label=False):
        self.sizes = size
        self.batch_size = batch_size
        self.onehot_label = onehot_label
        self.output_map = output_map
        super().__init__(pipelines=pipeline, size=size, auto_reset=auto_reset, output_map=output_map)

    def __next__(self):
        if self._first_batch is not None:
            batch = self._first_batch
            self._first_batch = None
            return batch
        data = super().__next__()[0]
        if self.onehot_label:
            return [data[self.output_map[0]], data[self.output_map[1]].squeeze().long()]
        else:
            return [data[self.output_map[0]], data[self.output_map[1]]]
    
    def __len__(self):
        if self.sizes%self.batch_size==0:
            return self.sizes//self.batch_size
        else:
            return self.sizes//self.batch_size+1
```

**Comment by JanuszL** [comment]
One thing is:
```
def __len__(self):
        if self.sizes%self.batch_size==0:
            return self.sizes//self.batch_size
        else:
            return self.sizes//self.batch_size+1
```
You are not calculating the size correctly, so that is why you get 3181. It should take into account the number of GPUs as well.
Another one is:
```
data = super().__next__()[0]
```
That is why you cannot do `train_loader._data_batches[1]`.
What I recommend is to use DALIGenericIterator directly.

**Comment by cs-heibao** [comment]
thanks, I will try it later. 发自我的华为手机-------- 原始邮件 --------发件人： Janusz Lisiecki <notifications@github.com>日期： 2020年7月28日周二 傍晚6:03收件人： NVIDIA/DALI <DALI@noreply.github.com>抄送： cs-heibao <Junjie_csu@163.com>, Author <author@noreply.github.com>主    题： Re: [NVIDIA/DALI] another problem about multi-gpu  (#2158)
One thing is:
def __len__(self):
        if self.sizes%self.batch_size==0:
            return self.sizes//self.batch_size
        else:
            return self.sizes//self.batch_size+1

You are not calculating the size correctly, so that is why you get 3181. It should take into account the number of GPUs as well.
Another one is:
data = super().__next__()[0]

That is why you cannot do train_loader._data_batches[1].
What I recommend is to use DALIGenericIterator directly.
